sql/alter_table.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/sql/privilege"
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// AlterTable creates a table.
+// Privileges: CREATE on database.
+//   Notes: postgres/mysql require CREATE on database.
+func (p *planner) AlterTable(n *parser.AlterTable) (planNode, error) {
+	if err := n.Table.NormalizeTableName(p.session.Database); err != nil {
+		return nil, err
+	}
+
+	dbDesc, err := p.getDatabaseDesc(n.Table.Database())
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if table exists.
+	tbKey := tableKey{dbDesc.ID, n.Table.Table()}.Key()
+	gr, err := p.txn.Get(tbKey)
+	if err != nil {
+		return nil, err
+	}
+	if !gr.Exists() {
+		if n.IfExists {
+			// Noop.
+			return &valuesNode{}, nil
+		}
+		// Key does not exist, but we want it to: error out.
+		return nil, fmt.Errorf("table %q does not exist", n.Table.Table())
+	}
+
+	tableDesc, err := p.getTableDesc(n.Table)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := p.checkPrivilege(tableDesc, privilege.CREATE); err != nil {
+		return nil, err
+	}
+
+	for _, cmd := range n.Cmds {
+		switch t := cmd.(type) {
+		case *parser.AlterTableAddColumn:
+			d := t.ColumnDef
+			col, idx, err := makeColumnDefDescs(d)
+			if err != nil {
+				return nil, err
+			}
+			tableDesc.Columns = append(tableDesc.Columns, *col)
+			if idx != nil {
+				if d.PrimaryKey {
```

Comment:
Like below, I don't think we'll ever be able to change the primary key after the table is created, so this can be simplified.